File: examples/quickstart.rs
//! Quickstart — Your first full-text search with Iris
//!
//! This minimal example shows how to:
//! 1. Create an in-memory search engine
//! 2. Define a schema with text fields
//! 3. Index documents
//! 4. Search with a simple term query
//!
//! Run with: `cargo run --example quickstart`

mod common;

use iris::lexical::{TermQuery, TextOption};
use iris::{Document, Engine, Schema, SearchRequestBuilder};

#[tokio::main]
async fn main() -> iris::Result<()> {
    println!("=== Iris Quickstart ===\n");

    // 1. Create storage and schema
    let storage = common::memory_storage()?;
    let schema = Schema::builder()
        .add_text_field("title", TextOption::default())
        .add_text_field("body", TextOption::default())
        .build();

    // 2. Create engine (no embedder needed for lexical-only search)
    let engine = Engine::new(storage, schema).await?;

    // 3. Index documents
    engine
        .add_document(
            "doc1",
            Document::builder()
                .add_text("title", "Introduction to Rust")
                .add_text("body", "Rust is a systems programming language focused on safety and performance.")
                .build(),
        )
        .await?;

    engine
        .add_document(
            "doc2",
            Document::builder()
                .add_text("title", "Python for Data Science")
                .add_text("body", "Python is a versatile language widely used in data science and machine learning.")
                .build(),
        )
        .await?;

    engine
        .add_document(
            "doc3",
            Document::builder()
                .add_text("title", "Web Development with JavaScript")
                .add_text("body", "JavaScript powers the modern web, from frontend frameworks to backend services.")
                .build(),
        )
        .await?;

    engine.commit().await?;
    println!("Indexed 3 documents.\n");

    // 4. Search for "rust"
    println!("[Search] term 'rust' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .with_lexical(Box::new(TermQuery::new("body", "rust")))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    // 5. Search for "language"
    println!("\n[Search] term 'language' in body:");
    let results = engine
        .search(
            SearchRequestBuilder::new()
                .with_lexical(Box::new(TermQuery::new("body", "language")))
                .limit(5)
                .build(),
        )
        .await?;
    common::print_search_results(&results);

    Ok(())
}

File: laurus-ruby/examples/quickstart.rb
# frozen_string_literal: true

# Quick start example for the laurus Ruby binding.
#
# Usage:
#   cd laurus-ruby
#   bundle exec rake compile
#   ruby -Ilib examples/quickstart.rb

require "laurus"

# Create a schema with two text fields.
schema = Laurus::Schema.new
schema.add_text_field("title")
schema.add_text_field("body")
schema.set_default_fields(%w[title body])

# Create an in-memory index.
index = Laurus::Index.new(schema: schema)

# Add documents.
index.add_document("doc1", { "title" => "Rust Programming", "body" => "Safety and speed." })
index.add_document("doc2", { "title" => "Python Basics", "body" => "Versatile language." })
index.commit

# Search with a DSL string.
puts "=== DSL search ==="
results = index.search("programming", limit: 5)
results.each do |r|
  puts "  #{r.id} (score: #{r.score}): #{r.document['title']}"
end

# Search with a TermQuery object.
puts "\n=== TermQuery search ==="
results = index.search(Laurus::TermQuery.new("body", "safety"), limit: 5)
results.each do |r|
  puts "  #{r.id} (score: #{r.score}): #{r.document['title']}"
end

# Index statistics.
puts "\n=== Stats ==="
stats = index.stats
puts "  Document count: #{stats['document_count']}"
